### PR TITLE
feat: configs for hparam optimization

### PR DESCRIPTION
**Pull request by hlee9212**
# What's Added?

This PR contains configurations used during the initial hyperparameter optimization. Please start with the file `configs/experiment/01_hyperparameter_optimization/starting_config.yaml`. 

<details>
<summary>Explanation of starting_config.yaml</summary>

```yaml
# @package _global_

# to execute this experiment run:
# python train.py experiment=example

defaults:
  - override /data: ycb_standard
  - override /model: base_vit_module
  - override /callbacks: default
  - override /trainer: default
  - _self_

# all parameters below will be merged with parameters from default configurations set above
# this allows you to overwrite only specified parameters

tags: ["ycb", "vit"]

seed: 12345

trainer:
  accelerator: mps
  max_epochs: 200
  num_sanity_val_steps: 0

model:
  compile: false # For debugging
  net:
    model_name: vit-b16-224-in21k
    freeze_backbone: false
  rotation_weight: 1.0

logger:
  wandb:
    tags: ${tags}
    group: "ycb"
    project: "benchmark_vit"
    name: "TRAIN_vit-b16-224-in21k"
  aim:
    experiment: "benchmark_vit"

hydra:
  run:
    dir: ${paths.log_dir}/${logger.wandb.name}
  job_logging:
    handlers:
      file:
        filename: ${paths.log_dir}/${logger.wandb.name}/${task_name}.log

callbacks:
  model_checkpoint:
    dirpath: ${paths.log_dir}/${logger.wandb.name}/checkpoints
```

First, the defaults section:
```yaml
defaults:
  - override /data: ycb_standard
  - override /model: base_vit_module
  - override /callbacks: default
  - override /trainer: default
  - _self_
```
These are overriding what we set in `train.yaml` (from PR #33). It's a bit redundant here since they are the same, except for the `_self_` at the end. This tells hydra to "import" or "inherit" all the above configs, and then incorporate all the specific parameters from below.

As a starting point for hyperparameter optimization, you can see I have set:
```yaml
trainer:
  accelerator: mps
  max_epochs: 200
  num_sanity_val_steps: 0
```
^ Train on `mps` (GPU for Apple Silicon) for up to 200 epochs. 

```yaml
model:
  compile: false # For debugging
  net:
    model_name: vit-b16-224-in21k
    freeze_backbone: false
  rotation_weight: 1.0
```
^ Use the ViT-B/16 model. 

The rest are just some WandB settings and path settings. 
</details>

---

Once you have read the `starting_config.yaml`, you can easily see how hyperparameter optimization was done. In the first experiment (`01_early_stopping`):

```yaml
# @package _global_

defaults:
  - /experiment/hyperparameter_optimization/starting_config
  - _self_

callbacks:
  early_stopping: null

logger:
  wandb:
    name: "vit-b16_no_early_stopping"
```

**How to read it:** Take all the settings from `starting_config.yaml`, but then (1) set early_stopping callback to none and (2) update name of wandb experiment (Optional, it just helps me find experiments in WandB). 

From then on, I kept adding to the config if the hyperparameter was helpful, e.g. kept gradient clipping to 1.0, etc. and setting the best value, e.g. learning rate after testing several different ones. 

The final config after optimization is at `01_hyperparameter_optimization/optimized_config.yaml`, which was used as the starting point for DMC experiments. 

**Note:** The hyperparameter experiments are not expected to be compatible with existing codebase - for example, after testing out different prediction head architectures and settling on one architecture, I deleted the argument for choosing different heads, i.e.

```yaml
# 05_multilayer_heads/multilayer_heads.yaml
model:
  net:
    classification_head_type: "simple"
    quaternion_head_type: "simple"
```
Trying to run this experiment won't work because I removed the arguments `classification_head_type` and `quaternion_head_type` in our ViT class. This helps keep the code clear and easy to read (it was much longer before due to lots of arguments for flags or setting head types, etc.). For transparencyh, the results of hyperparameter optimization are recorded in README, which I have shared in Slack before. 😄 

Of course, all other experiments are compatible because they produce results for figures in DMC paper. :)

### CODE REVIEW

File: pytorch/configs/experiment/01_hyperparameter_optimization/13_baseline_models/vit_b16.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# @package _global_
+# Configuration for testing NormLinearHead (LayerNorm + Linear)
```
Comment (nielsleadholm) [inline pytorch/configs/experiment/01_hyperparameter_optimization/13_baseline_models/vit_b16.yaml:2]: **nitpick:** It looks like all the yaml files contain this comment, it would be good to clean these up.

File: pytorch/configs/experiment/01_hyperparameter_optimization/README.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    # Copy RGB weights and initialize depth channel
+    with torch.no_grad():
+        depth_weight = rgb_weights.mean(dim=1, keepdim=True)  # (out_c, 1, k, k)
+        new_weight = torch.cat(
+            [rgb_weights, depth_weight], dim=1
+        )
+```
+
+- Dataset and DataLoaders:
+  - Dataset contains 1,078 images (77 objects x 14 rotations), randomly split into train/val with 80/20 ratio
```
Comment (nielsleadholm) [inline pytorch/configs/experiment/01_hyperparameter_optimization/README.md:30]:
```suggestion
  - Dataset contains 1,078 images (77 objects x 14 rotations), randomly split into train/val with 80/20 ratio during hyperparameter optimization
```

File: pytorch/configs/experiment/01_hyperparameter_optimization/README.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        T.RandomResizedCrop(224, scale=(0.9, 1.0), ratio=(0.9, 1.1)),
+        T.ColorJitter(0.2, 0.2, 0.2, 0.05),
+        T.Normalize(mean=[0.5, 0.5, 0.5], std=[0.5, 0.5, 0.5]),
+    ]
+)
+```
+
+- **Results**:
+  - With augmentation: `val/class_acc` = 87.5%; `test/class_acc` = 78.4%
+- **Final Decision**: Omit data augmentation to match training conditions with other models
```
Comment (nielsleadholm) [inline pytorch/configs/experiment/01_hyperparameter_optimization/README.md:88]:
```suggestion
- **Final Decision**: Omit data augmentation to match training conditions with Monty models (receive no data-augmentation, and are restricted from using their optimal exploration policies that would acquire more data)
```

File: pytorch/configs/experiment/01_hyperparameter_optimization/README.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+### 9. Loss Weight Tuning ($\\lambda\_\\text{rot}$)
+
+- **Reason**: Validate empirical choice of $\\lambda\_\\text{rot} = 1.0$
+- **Results**:
+  - $\\lambda\_\\text{rot} = 0.1$: `val/class_acc = 92.6%`, `val/rotation_error = 113.7°`
+  - $\\lambda\_\\text{rot} = 1.0$: `val/class_acc = 90.3%`, `val/rotation_error = 109.2°`
+  - $\\lambda\_\\text{rot} = 3.0$: `val/class_acc = 89.4%`, `val/rotation_error = 106.2°`
+  - $\\lambda\_\\text{rot} = 5.0$: `val/class_acc = 88.9%`, `val/rotation_error = 105.8°`
+- **Decision**: Keep $\\lambda\_\\text{rot} = 1.0$
```
Comment (nielsleadholm) [inline pytorch/configs/experiment/01_hyperparameter_optimization/README.md:128]:
```suggestion
- **Decision**: Keep $\\lambda\_\\text{rot} = 1.0$ to balance accuracy and rotation error performance.
```

**Review by nielsleadholm — APPROVED** [review]
Thanks for the thorough documentation as always Hojae. Looks great, just added a couple small comment suggestions for before you merge.
